### ISSUE TITLE
python tests executed on source instead of artifacts

### ISSUE THREAD

**Issue by DavHau** [NixOS/nixpkgs#123561]
**Describe the bug**
The way python tests are executed in nixpkgs, it is likely that only the source is tested, not the build artifacts. Those are not necessarily the same.

The reason for this is that we execute our tests from the source directory (since it contains the `./tests/` dir).
But this directory usually also contains the main modules directory which has the same name as the module in $out.
Therefore the tests executed via `setup.py test`, `pytest` or  `nose` will actually import the python module from the unpacked source, not the one from $out.

Usually the files are the same and it is not an issue, but there is no guarantee for it. I was just wasting hours wondering why the extension modules of a python package were not found during the tests. They existed in $out, they just didn't exist in the source.

I think this issue is critical regarding reliability. Theoretically even if $out/lib would be empty most of our current tests would still succeed (I guess).

I think it would be good if we could come up with a generic solution for that, so that maintainers do not have to think about this.

We could: 
- patch the python interpreter to not prepend the current directory to  `sys.path`
- make sure to delete the relevant files from the source directory before starting tests (probably not trivial)
- Write our check hooks in a way so they will use another root dir. (But we don't even have hooks for some testing tools).

**Notify maintainers**
@FRidh @jonringer

**Comment by jonringer** [comment]
For packages which have compiled extensions (where I think testing in $out makes the most sense), I think I have to include logic to prepend `$out/${python.sitePackages}` to `PYTHONPATH`. We should probably do something similar for the testing hooks, and expose an option to disable it.

**Comment by DavHau** [comment]
`$out/${python.sitePackages}` is currently already the first entry of PYTHONPATH. But still the python interpreter always prepends the current directory to `sys.path`. Maybe we can configure/wrap the interpreter to not do that.

**Comment by DavHau** [comment]
In case you need a package to test this behavior, see https://github.com/NixOS/nixpkgs/pull/123553/. If you remove the `preCheck`, the tests will fail.

**Comment by FRidh** [comment]
Yes, this is a real issue, but unfortunately there is no reliable way to do
this during a build other than patching the interpreter.

https://discuss.python.org/t/python-flag-envvar-not-to-put-current-directory-to-sys-path-but-dont-ignore-pythonpath/4235

**Comment by DavHau** [comment]
I guess this issue is one more reason for separating build and check phase into different derivations.

But for the meantime, how about a `prepareForTests` hook that reads a list `keepDirsForCheckPhase` and deletes all directories from the current directory which are not listed?
By default this list would include `./test ./tests`.

**Comment by FRidh** [comment]
Yes we can try and test it on all our packages.

**Comment by veprbl** [comment]
Why `preCheck = "pushd $out/${python.sitePackages}"` doesn't work for us?

**Comment by DavHau** [comment]
> Why `preCheck = "pushd $out/${python.sitePackages}"` doesn't work for us?

The problem with this is that it makes using testing tools like pytest/nose more complicated. Since the ./tests directory is not anymore in the current directory, you'd need to tell pytest the path to the `./tests` dir via flag. The default behavior of pytest then is to switch back to the original directory and execute the tests from there. So you're back to the original problem. You would also need to set `--rootdir`.
If we go this way we will rely on the users all knowing about this and calling the tests the right way. We could include that into the test hooks, but they are not always used.

**Comment by DavHau** [comment]
> But for the meantime, how about a `prepareForTests` hook that reads a list `keepDirsForCheckPhase` and deletes all directories from the current directory which are not listed?
> By default this list would include `./test ./tests`.

I thought it would probably be easier to do that if we first got rid of `setup.py` invoked tests. `setup.py` often reads arbitrary other files which would make it more complicated to find and list them.
I have issued a PR here to deprecate `setuptoolsCheckHook`: https://github.com/NixOS/nixpkgs/pull/124002
